public/app/features/explore/state/query.ts

```diff
@@ -572,59 +607,72 @@ export const runQueries = (
       if (live) {
         for (const type of SUPPLEMENTARY_QUERY_TYPES) {
           dispatch(
-            storeSupplementaryQueryDataProviderAction({
+            cleanSupplementaryQueryDataProviderAction({
               exploreId,
-              dataProvider: undefined,
               type,
             })
           );
           dispatch(cleanSupplementaryQueryAction({ exploreId, type }));
         }
-
-        // In this whole part., we need to figure out
-        // checking the type of enabled supp queries
-        // then for which enabled supp queries has data source support
-        // and then we need to run the supp queries
-        // but we need to make sure that supp queries that dont work
-        // return undefined provider
-        // we should also make sure we store the type of provider that
-        // was last stored
-      } else if (hasLogsVolumeSupport(datasourceInstance)) {
-        // we always prepare the logsVolumeProvider,
-        // but we only load it, if the logs-volume-histogram is enabled.
-        // (we need to have the logsVolumeProvider always actual,
-        // even when the visuals are disabled, because when the user
-        // enables the visuals again, we need to load the histogram,
-        // so we need the provider)
-        const sourceRequest = {
-          ...transaction.request,
-          requestId: transaction.request.requestId + '_log_volume',
-        };
-        const type = SupplementaryQueryType.LogsVolume;
-        const dataProvider = datasourceInstance.getLogsVolumeDataProvider(sourceRequest);
-        dispatch(
-          storeSupplementaryQueryDataProviderAction({
+        // We support multiple supplementary queries, so we check if data source supports at least 1
+        // If it does, we then check for each of them individually and process accordingly
+      } else if (hasSupplementaryQuerySupport(datasourceInstance)) {
+        if (hasLogsVolumeSupport(datasourceInstance)) {
```

Comment:
Is it possible to make it more generic so we don't have `if` for each type, but rather iterate over all of them?

Reply:
Good idea, I refactored it in here https://github.com/grafana/grafana/pull/61298/commits/e9a6a1d3c66b4d5856ea833345ace1afea4e8706. Let me know what do you think!

Reply:
Good idea! I have refactored it in https://github.com/grafana/grafana/pull/61298/commits/e9a6a1d3c66b4d5856ea833345ace1afea4e8706. (I removed redundant Object.values when looping over supplementaryQueriesList in https://github.com/grafana/grafana/pull/61298/commits/41abe6411b9b44a77526ab713d04be7d492785b7. I forgot to fix it in first commit.) And then I fixed additional bug in logic in https://github.com/grafana/grafana/pull/61298/commits/b31a65fabed8c3412ab82b647a1a7584425651f8.

Let me know what do you think